api/handlers/iq.py

```diff
@@ -68,33 +68,84 @@ async def get_iq_data(
         raise HTTPException(status_code=400, detail=str(e))
 
 
-async def get_byte_stream(
+async def calculate_iq_data(
+    block_indexes,
+    block_size,
+    format,
+    iq_file,
+    azure_client,
+):
+    iq_data_list = await get_byte_streams(
+        block_indexes,
+        block_size,
+        format,
+        iq_file,
+        azure_client,
+    )
+    for iq_data in iq_data_list:
+        yield iq_data
+
+
+async def get_byte_streams(
     block_indexes, block_size, bytes_per_iq_sample, iq_file, azure_client
 ):
+    chunk_size = 200 * 1024 // block_size
     block_indexes_arrs = find_smallest_and_largest_next_to_each_other(block_indexes)
 
-    for block_index_arr in block_indexes_arrs:
-        offsetBytes = block_index_arr[0] * block_size * bytes_per_iq_sample
-        countBytes = (
-            (block_index_arr[1] - block_index_arr[0] + 1)
-            * block_size
-            * bytes_per_iq_sample
-        )
+    block_indexes_chunks = []
+    for i in block_indexes_arrs:
+        if i[1] - i[0] > chunk_size:
+            indexes = [
+                [j, min(j + chunk_size - 1, i[1])]
+                for j in range(i[0], i[1] + 1, chunk_size)
+            ]
+            block_indexes_chunks.extend(indexes)
+        else:
+            block_indexes_chunks.append(i)
 
-        blob_size = await azure_client.get_file_length(iq_file)
-        if blob_size < offsetBytes:
-            return
-        if blob_size < offsetBytes + countBytes:
-            countBytes = blob_size - offsetBytes
-        st = time.time()
-        content = await azure_client.get_blob_content(
-            filepath=iq_file, offset=offsetBytes, length=countBytes
-        )
-        print(
-            f"get_blob_content {iq_file}/{offsetBytes}/{countBytes} took {time.time() - st} seconds"
+    blob_size = await azure_client.get_file_length(iq_file)
+
+    tasks = [
+        get_byte_stream(
+            block_index_chunk,
+            block_size,
+            bytes_per_iq_sample,
+            iq_file,
+            azure_client,
+            blob_size,
         )
+        for block_index_chunk in block_indexes_chunks
+    ]
+    return await asyncio.gather(*tasks)
```

Comment:
Since this does not use pools wouldn't this be actually sending potentially 800 requests at the same time to the storage account? Would this get us easily to be throttle because Of too many concurrent requests?

Reply:
We could use semaphore to limit requests:
https://stackoverflow.com/questions/48483348/how-to-limit-concurrency-with-python-asyncio

Reply:
I think that could work defintly!

Reply:
added semaphore to limit requests